container/build.py

```diff
@@ -99,32 +108,89 @@
+assert set(SUPPORTED_TYPES) \
         == set(TYPE_PACKAGE_MAP.keys()) \
         == set(TYPE_TARGET_MAP.keys()) \
         == set(TYPE_TARBALL_MAP.keys()), \
             "TYPES ARE OUT OF SYNC"
 
 
-def main(type_, project, container, tag, async_, bucket, local):
+def main(type_, project, container, tag, async_, bucket, local, bazel_version, map = None):
   '''Runs the build. More info in module docstring at the top.
   '''
+  type_package_map = TYPE_PACKAGE_MAP
+  type_target_map = TYPE_TARGET_MAP
+  type_tarball_map = TYPE_TARBALL_MAP
+
+  if map: # Override the map values
+    try:
+      map_module = imp.load_source("map", map)
+    except IOError as e:
+      print("Error reading map file.\n" ,e)
+    
+    try:
+      type_package_map = map_module.TYPE_PACKAGE_MAP
+      type_target_map = map_module.TYPE_TARGET_MAP
+      type_tarball_map = map_module.TYPE_TARBALL_MAP
+    except AttributeError as e:
+      print("Error getting attributes from map file.\n", e)
+
+  # Gets the project root (for calling bazel targets)
   project_root = subprocess.check_output(
       shlex.split("git rev-parse --show-toplevel")).strip()
-  package = TYPE_PACKAGE_MAP[type_]
-  target = TYPE_TARGET_MAP[type_]
-  tarball = TYPE_TARBALL_MAP[type_]
-
+  package = type_package_map[type_]
+  target = type_target_map[type_]
+  tarball = None
+  if bucket:
+    tarball = type_tarball_map[type_]
+
+  # Gets the base directory of the bazel-toolchains repo (relative to project_root)
+  # This is for referencing yaml files and mounting the project to gcloud.
+  # Allows the bazel-toolchains repo to be cloned in any subdirectory
+  # of another project and bazel_toolchains_base_dir will store the relative path
+  # from the root of that git project to the root of the bazel-toolchains git project.
+  # Ex. If we are in folder foo and the structure looks like this:
+  # foo/
+  #     subdir/
+  #         bazel-toolchains
+  #     ...
+  # 
+  # Then if we call build.py with the terminal running in foo,
+  # bazel_toolchains_base_dir == "subdir/bazel-toolchains"
+  # (This also allows for renaming of the bazel-toolchains folder as the variable
+  # will store the path regardless of the directory names) 
+  os.chdir(os.path.dirname(__file__))
+  bazel_toolchains_base_dir = os.path.relpath(subprocess.check_output(
+      shlex.split("git rev-parse --show-toplevel")).strip(), project_root)
+  
   # We need to start the build from the root of the project, so that we can
   # mount the full root directory (to use bazel builder properly).
   os.chdir(project_root)
-  # We need to run clean to make sure we don't mount local build outputs
-  subprocess.call(["bazel", "clean"])
+
+  # Ensure all BUILD files under /third_party have the right permission.
+  # This is because in some systems the BUILD files under /third_party
+  # (after git clone) will be with permission 640 and the build will
+  # fail in Container Builder.
+  for dirpath, _, files in os.walk("third_party"):
```

Comment:
This should stay in cloud_build function.

Reply:
Moved it back